### PR TITLE
Fix implied formats, file watching, new source file creating during edits

### LINKED ISSUE #50086: nodenext doesn't pick up on addition of `type` field to `package.json`
Here's a quick way you can clone this repro locally

```sh
https://github.com/DanielRosenwasser/daniel-ts-repros/
cd daniel-ts-repros
git checkout nodenext-package-watch-bug
```

______

```ts
// src/fileA.ts
import { foo } from "./fileB.mjs";

foo();
```

```ts
// src/fileB.mts
export function foo() {
}
```

```json5
// src/tsconfig.json
{
  "compilerOptions": {
    "target": "es2016",
    "module": "Node16",
    "outDir": "../out"
  }
}
```

```json5
// package.json

{
  "name": "nodenext-package-watch-bug",
  "version": "1.0.0",
  "description": "Repro",
  "author": "Daniel"
}
```

Currently this produces the following result in the editor or in `--watch` mode:

```
Module './fileB.mjs' cannot be imported using this construct. The specifier only resolves to an ES module, which cannot be imported synchronously. Use dynamic import instead.ts
```

If you add a `"type": "module"` to the `package.json`, then the error will not go away until you restart `tsc` or TS Server.

```diff
-  "author": "Daniel"
+  "author": "Daniel",
+  "type": "module"
```

### PR DESCRIPTION
**Pull request by sheetalkamat**
This change includes:
1. When program is created, we track package.json per file that affect its implied format
2. We watch those files to invalidate resolutions of that file
3. If oldSourceFile.impliedNodeFormat is not same as new one, we need to create new file, so we pass that information to host
4. Document registry handles implied node format as part of its key to get the file and release it 
5. Package.json files are being watched so we know when to clear the package json info cache.
6. Fixed issue with module resolution cache not passed around in watch mode
7. If implied file format is determined by package.json persence/absence and file is module, its reported with `--explainFiles`

Fixes #50086

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@typescript-bot pack this

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -926,12 +937,17 @@ namespace ts {
...
         };
 
         function clear() {
-            preDirectoryResolutionCache.clear();
+            clearAllExceptPackageJsonInfoCache();
             packageJsonInfoCache!.clear();
         }
+
+        function clearAllExceptPackageJsonInfoCache() {
+            preDirectoryResolutionCache.clear();
```
Comment (weswigham) [inline src/compiler/moduleNameResolver.ts:949]: This one is also `pre` instead of `per` :)

File: src/compiler/watch.ts [hunk 2]
```diff
@@ -224,26 +224,59 @@ namespace ts {
...
-            (result ||= []).push(chainDiagnosticMessages(
+            (result ??= []).push(chainDiagnosticMessages(
                 /*details*/ undefined,
                 Diagnostics.File_redirects_to_file_0,
                 toFileName(file.redirectInfo.redirectTarget, fileNameConvertor)
             ));
         }
+        if (isExternalOrCommonJsModule(file)) {
+            switch (file.impliedNodeFormat) {
+                case ModuleKind.ESNext:
```
Comment (weswigham) [inline src/compiler/watch.ts:252]: Should both of these have a branch for when the file extension sets the format, for `.cts`, `.cjs`, `.mjs` and `.mts` files?

Comment (sheetalkamat) [reply src/compiler/watch.ts:252]: @DanielRosenwasser and myself discussed offline before this change and decided to only spit this out when package.json determines the context.

**Review by weswigham — APPROVED** [review]

**Pull request merged**